core/src/main/java/com/datastrato/gravitino/authorization/PermissionManager.java

```diff
@@ -0,0 +1,370 @@
+      String metalake, String role, String user, Instant latestDropRoleInstant) {
+    try {
+      AuthorizationUtils.checkMetalakeExists(store, metalake);
+
+      checkRoleExists(metalake, role);
+
+      RoleEntity roleEntity =
+          store.get(
+              NameIdentifierUtils.ofRole(metalake, role), Entity.EntityType.ROLE, RoleEntity.class);
+
+      AtomicBoolean removed = new AtomicBoolean(true);
+
+      store.update(
+          NameIdentifierUtils.ofUser(metalake, user),
+          UserEntity.class,
+          Entity.EntityType.USER,
+          userEntity -> {
+            RoleInformation roleInformation =
+                getValidRolesInformation(
+                    metalake,
+                    latestDropRoleInstant,
+                    userEntity.roles(),
+                    userEntity.roleIds(),
+                    userEntity.latestGcRolesTime());
+
+            roleInformation.roleNames.remove(roleEntity.name());
+            removed.set(roleInformation.roleIds.remove(roleEntity.id()));
+            if (!removed.get()) {
+              LOG.warn("Role {} doesn't exist in the user {} of metalake {}", role, user, metalake);
+            }
+
+            AuditInfo auditInfo =
+                AuditInfo.builder()
+                    .withCreator(userEntity.auditInfo().creator())
+                    .withCreateTime(userEntity.auditInfo().createTime())
+                    .withLastModifier(PrincipalUtils.getCurrentPrincipal().getName())
+                    .withLastModifiedTime(Instant.now())
+                    .build();
+            return UserEntity.builder()
+                .withId(userEntity.id())
+                .withNamespace(userEntity.namespace())
+                .withName(userEntity.name())
+                .withLatestGcRolesTime(Instant.now())
+                .withRoleNames(roleInformation.roleNames)
+                .withRoleIds(roleInformation.roleIds)
+                .withAuditInfo(auditInfo)
+                .build();
+          });
+      return removed.get();
+    } catch (NoSuchEntityException nse) {
+      LOG.warn("User {} does not exist in the metalake {}", user, metalake, nse);
+      throw new NoSuchUserException(USER_DOES_NOT_EXIST_MSG, user, metalake);
+    } catch (IOException ioe) {
+      LOG.error(
+          "Revoking role {} from  user {} failed in the metalake {} due to storage issues",
+          role,
+          user,
+          metalake,
+          ioe);
+      throw new RuntimeException(ioe);
+    }
+  }
+
+  private RoleInformation getValidRolesInformation(
+      String metalake,
+      Instant latestDropRoleInstant,
+      List<String> roleNames,
+      List<Long> roleIds,
+      Instant latestGcInstant) {
+
+    if (latestDropRoleInstant.isAfter(latestGcInstant)) {
+      List<RoleEntity> roleEntities = removeInvalidRoles(metalake, roleNames);
+      roleNames = Lists.newArrayList(toRoleNames(roleEntities));
+      roleIds = Lists.newArrayList(toRoleIds(roleEntities));
+    }
```

Comment:
Does it really solve the problem? It's unrelated to GC, why do you call it GcXXX?

Reply:
I think it's a kind of GC that we remove invalid roles. If it will make people confusing, I will rename it.